contracts/peripheral/YieldStreamer.sol

```diff
@@ -0,0 +1,466 @@
+            // Remove id_ from donor's depositId array
+            if (userIndices[i] == id_) {
+                userIndices[i] = userIndices[userIndices.length - 1]; // Delete integer from array by swapping with last element and calling pop()
+                userIndices.pop();
+                break;
+            }
+        }
+
+        IERC20(sOHM).safeTransfer(msg.sender, totalSOhm);
+        IERC20(DAI).safeTransfer(msg.sender, totalDai);
+
+        emit Withdrawn(msg.sender, totalSOhm);
+    }
+
+    /**
+        @notice User updates the minimum amount of DAI threshold before upkeep sends DAI to recipients wallet
+        @param id_ Id of the deposit
+        @param threshold_ amount of DAI
+    */
+    function updateUserMinDaiThreshold(uint256 id_, uint256 threshold_) external override {
+        require(threshold_ >= minimumDaiThreshold, "minimumDaiThreshold too low");
+
+        DepositInfo storage userDeposit = depositInfo[id_];
+        require(userDeposit.depositor == msg.sender, "Deposit is not yours");
+
+        userDeposit.userMinimumDaiThreshold = threshold_;
+    }
+
+    /**
+        @notice User updates the minimum amount of time passes before the deposit is included in upkeep
+        @param id_ Id of the deposit
+        @param paymentInterval_ amount of time in seconds
+    */
+    function updatePaymentInterval(uint256 id_, uint256 paymentInterval_) external override {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        require(userDeposit.depositor == msg.sender, "Deposit is not yours");
+
+        userDeposit.paymentInterval = paymentInterval_;
+    }
+
+    /**
+        @notice Upkeeps all deposits if they are eligible to be upkept. Converts excess yield from sOHM to DAI. Sends the yield to recipient wallets if above user set threshold.
+    */
+    function upkeep() external override {
+        uint256 totalOhm;
+
+        for (uint256 i = 0; i < activeDepositIds.length; i++) {
+            uint256 currentId = activeDepositIds[i];
+
+            if (isUpKeepEligible(currentId)) {
+                DepositInfo storage currentDeposit = depositInfo[currentId];
+                totalOhm += _getOutstandingYield(currentDeposit.principalAmount, currentDeposit.agnosticAmount);
+                currentDeposit.lastUpkeepTimestamp = block.timestamp;
+            }
+        }
+
+        uint256 feeToDao = (totalOhm * feeToDaoPercent) / 1000;
+        IERC20(sOHM).safeTransfer(authority.governor(), feeToDao);
+        uint256 totalOhmToSwap = totalOhm - feeToDao;
+        staking.unstake(address(this), totalOhmToSwap, false, false);
+
+        require(IERC20(OHM).approve(address(sushiRouter), totalOhmToSwap), "approve failed");
+        uint256[] memory calculatedAmounts = sushiRouter.getAmountsOut(totalOhmToSwap, sushiRouterPath);
+        uint256[] memory amounts = sushiRouter.swapExactTokensForTokens(
+            totalOhmToSwap,
+            (calculatedAmounts[1] * (1000 - maxSwapSlippagePercent)) / 1000,
```

Comment:
Do you think the way we are calculating the max slippage here could leave this open to a sandwich attack? Take a look at the Uniswap V2 [docs](https://docs.uniswap.org/protocol/V2/guides/smart-contract-integration/trading-from-a-smart-contract)

> When trading from a smart contract, the most important thing to keep in mind is that access to an external price source is required. Without this, trades can be frontrun for considerable loss.

Maybe something like this:
* Flashloan a bunch of ETH
* Buy DAI via ETH->OHM-DAI, pushing the price of OHM down
* run `upkeep`, selling OHM for a small amount of DAI
* buyback ETH via DAI->OHM-ETH claiming the amount of DAI that should have been swapped for here.

My guess is the OHM-DAI pool has too much liquidity and these amounts will be too small to make this profitable.

Reply:
Yea so from our discussions either the amount would be too small for MEV attacks. Or if that is not the case route the transaction through flashbots rpc. Haven't looked into that yet just took Indigos word for it. Just assumed there would be a way to avoid getting attacked so i've put it on the backfoot for now.